src/messaging/ExchangeContext.cpp

```diff
@@ -635,6 +645,15 @@ void ExchangeContext::MessageHandled()
 #if CONFIG_DEVICE_LAYER && CHIP_DEVICE_CONFIG_ENABLE_SED
     UpdateSEDIntervalMode();
 #endif
+#if CONFIG_DEVICE_LAYER
+    DeviceLayer::ChipDeviceEvent event;
+    event.Type        = DeviceLayer::DeviceEventType::kChipMsgReceivedEvent;
```

Comment:
How do we know this was a message received?  It might have been a message sent.  Or a response wait timeout.

Reply:
Thanks for clarifying the `MessageHandled`. I am moving this to `HandleMessage`. What is your opinion on this? Should we still notify the `ICDManager` for unexpected messages?

Reply:
Realistically, I think this is generally the wrong model.  The right model is that we should be in active mode while an exchange is waiting for a response, no?  So the right notifications are:

1. We sent a message, waiting for a response (kicks us into active mode, where we need to stay until the next notification)
2. No longer waiting for a response (timeout or response received).  This is where we can release the "lock" on active mode, but stay in it until the timer is hit.

Reply:
I guess if we send a message that does _not_ expect a response we should also do the threshold thing.

So if we want the ICDManager to implement the logic, the needed notifications are:

1. Message sent, no response expected.
2. Message sent, response expected; needs to include some way to identify _which message_ for purposes of notification 3.  Could be the message counter.
3. No longer waiting for a response for message (again, needs to include some way to indicate which message).

But I suspect in practice the correlation of messages and responses should just happen in the exchange itself, and the notifications should be:

1. Message sent, no response expected.
2. Message sent, waiting for a response.
3. No longer waiting for a response that we previously notified about.

Reply:
Since there can be multiple ExchangeContext Sending the notification for 
- Message sent, waiting for a response.
- No longer waiting for a response that we previously notified about.
I need a way to identify/Match the notification that clears the wait. Or maybe a count of how many responses notifications I am expecting

Reply:
Right, count is enough.  But you do need a count.

Reply:
Updated the eventing and "locking" in active mode as we discussed.